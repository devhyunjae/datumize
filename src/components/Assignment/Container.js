import styled from 'styled-components'

export default styled.div`
  margin-top: 15px;
  margin-bottom: 15px;
  width: 100%;
`
