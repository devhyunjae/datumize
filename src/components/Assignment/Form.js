import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Card, Button, Toaster, Position } from '@blueprintjs/core'

import UserDropDown from './UsersDropdown'
import ProjectsDropDown from './ProjectsDropdown'
import RolesDropDown from './RolesDropdown'
import * as actions from '../../actions/users'

const Container = styled.div`
  display: flex;
  flex: 1;
  justify-content: center;
  align-items: flex-start;
  flex-direction: column;
`

class Form extends PureComponent {
  static propTypes = {
    updateUser: PropTypes.func.isRequired,
    successMessage: PropTypes.string,
  }

  static defaultProps = {
    successMessage: null,
  }

  state = {
    selectedUser: null,
    selectedProject: null,
    selectedRole: null,
  }

  componentDidUpdate(prevProps) {
    const { successMessage } = this.props

    if (successMessage && (prevProps.successMessage !== successMessage)) {
      this.toaster.show({ message: successMessage })
    }
  }

  onClick = () => {
    const { selectedUser, selectedProject, selectedRole } = this.state
    this.props.updateUser({ user: selectedUser, project: selectedProject, role: selectedRole })
  }

  setSelectedUser = (selectedUser) => {
    this.setState({
      selectedUser,
    })
  }

  setSelectedProject = (selectedProject) => {
    this.setState({
      selectedProject,
    })
  }

  setSelectedRole = (selectedRole) => {
    this.setState({
      selectedRole,
    })
  }

  toasterHandler = (ref) => {
    this.toaster = ref
  }

  render() {
    return (
      <Card elevation={Card.ELEVATION_TWO}>
        <Container>
          <UserDropDown setSelectedUser={this.setSelectedUser} />
          <ProjectsDropDown setSelectedProject={this.setSelectedProject} />
          <RolesDropDown setSelectedRole={this.setSelectedRole} />
          <Button type="button" onClick={this.onClick} className="pt-fill">
            Save
          </Button>
        </Container>
        <Toaster position={Position.TOP} ref={this.toasterHandler} />
      </Card>
    )
  }
}

const mapStateToProps = state => ({
  successMessage: state.users.success,
})

export default connect(
  mapStateToProps,
  actions,
)(Form)
