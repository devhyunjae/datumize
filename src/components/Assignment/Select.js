import React from 'react'
import PropTypes from 'prop-types'

const Select = ({ children }) => (
  <div className="pt-select pt-fill">
    { children }
  </div>
)

Select.propTypes = {
  children: PropTypes.element.isRequired,
}

export default Select
