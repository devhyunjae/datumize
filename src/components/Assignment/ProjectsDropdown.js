import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import * as actions from '../../actions/projects'
import Container from './Container'
import Select from './Select'

class ProjectsDropdown extends PureComponent {
  static propTypes = {
    fetchProjects: PropTypes.func.isRequired,
    projects: PropTypes.array.isRequired,
    setSelectedProject: PropTypes.func,
  }

  static defaultProps = {
    setSelectedProject: () => {},
  }

  componentDidMount() {
    const { fetchProjects } = this.props
    fetchProjects()
  }

  componentDidUpdate() {
    const { projects, setSelectedProject } = this.props
    if (projects.length > 0) {
      const selectedProject = projects[0]
      setSelectedProject(selectedProject)
    }
  }

  changeEvent = (event) => {
    const { projects, setSelectedProject } = this.props
    const selectedProject = projects.find(project => project.id === parseInt(event.target.value, 0))
    setSelectedProject(selectedProject)
  }

  renderOptions() {
    const { projects } = this.props

    return projects.map(project => (
      <option value={project.id} key={project.id}>
        {project.name}
      </option>
    ))
  }

  render() {
    return (
      <Container>
        <h5>2. Choose a project</h5>
        <Select>
          <select onChange={this.changeEvent}>
            { this.renderOptions() }
          </select>
        </Select>
      </Container>
    )
  }
}

const mapStateToProps = state => ({
  projects: state.projects.projects,
})

export { ProjectsDropdown }

export default connect(
  mapStateToProps,
  actions,
)(ProjectsDropdown)

