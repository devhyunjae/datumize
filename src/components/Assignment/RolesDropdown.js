import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import * as actions from '../../actions/roles'
import Container from './Container'
import Select from './Select'

class RolesDropdown extends PureComponent {
  static propTypes = {
    fetchRoles: PropTypes.func.isRequired,
    roles: PropTypes.array.isRequired,
    setSelectedRole: PropTypes.func,
  }

  static defaultProps = {
    setSelectedRole: () => {},
  }

  componentDidMount() {
    const { fetchRoles } = this.props
    fetchRoles()
  }

  componentDidUpdate() {
    const { roles, setSelectedRole } = this.props
    if (roles.length > 0) {
      const selectedRole = roles[0]
      setSelectedRole(selectedRole)
    }
  }

  changeEvent = (event) => {
    const { roles, setSelectedRole } = this.props
    const selectedRole = roles.find(role => role.id === parseInt(event.target.value, 0))
    setSelectedRole(selectedRole)
  }

  renderOptions() {
    const { roles } = this.props

    return roles.map(role => (
      <option value={role.id} key={role.id}>
        {role.name}
      </option>
    ))
  }

  render() {
    return (
      <Container>
        <h5>3. Choose a role</h5>
        <Select>
          <select onChange={this.changeEvent}>
            { this.renderOptions() }
          </select>
        </Select>
      </Container>
    )
  }
}

const mapStateToProps = state => ({
  roles: state.roles.roles,
})

export { RolesDropdown }

export default connect(
  mapStateToProps,
  actions,
)(RolesDropdown)

