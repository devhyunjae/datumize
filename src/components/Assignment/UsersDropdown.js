import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import * as actions from '../../actions/users'
import Container from './Container'
import Select from './Select'

class UsersDropdown extends PureComponent {
  static propTypes = {
    fetchUsers: PropTypes.func.isRequired,
    users: PropTypes.array.isRequired,
    setSelectedUser: PropTypes.func,
  }

  static defaultProps = {
    setSelectedUser: () => {},
  }

  componentDidMount() {
    const { fetchUsers } = this.props
    fetchUsers()
  }

  componentDidUpdate() {
    const { users, setSelectedUser } = this.props
    if (users.length > 0) {
      const selectedUser = users[0]
      setSelectedUser(selectedUser)
    }
  }

  changeEvent = (event) => {
    const { users, setSelectedUser } = this.props
    const selectedUser = users.find(user => user.id === parseInt(event.target.value, 0))
    setSelectedUser(selectedUser)
  }

  renderOptions() {
    const { users } = this.props

    return users.map(user => (
      <option value={user.id} key={user.id} name={user.name}>
        {user.name}
      </option>
    ))
  }

  render() {
    return (
      <Container>
        <h5>1. Choose a user</h5>
        <Select>
          <select onChange={this.changeEvent}>
            { this.renderOptions() }
          </select>
        </Select>
      </Container>
    )
  }
}

const mapStateToProps = state => ({
  users: state.users.users,
})

export { UsersDropdown }

export default connect(
  mapStateToProps,
  actions,
)(UsersDropdown)

