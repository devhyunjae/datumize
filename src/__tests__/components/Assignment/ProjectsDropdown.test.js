import React from 'react'
import renderer from 'react-test-renderer'
import { ProjectsDropdown } from '../../../components/Assignment/ProjectsDropdown'

describe('<ProjectsDropdown>', () => {
  it('renders correctly', () => {
    const fetchedProjects = [
      { id: 1, name: 'project 1' },
      { id: 2, name: 'project 2' },
    ]
    const component = renderer.create(
      <ProjectsDropdown
        fetchProjects={() => {}}
        projects={fetchedProjects}
      />,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
