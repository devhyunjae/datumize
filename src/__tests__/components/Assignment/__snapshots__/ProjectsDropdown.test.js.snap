// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectsDropdown> renders correctly 1`] = `
<div
  className="sc-bdVaJa ePGVPP"
>
  <h5>
    2. Choose a project
  </h5>
  <div
    className="pt-select pt-fill"
  >
    <select
      onChange={[Function]}
    >
      <option
        value={1}
      >
        project 1
      </option>
      <option
        value={2}
      >
        project 2
      </option>
    </select>
  </div>
</div>
`;
